S3 upload support

pjanik opened the pull request:
This PR adds support of S3 upload to Glossary authoring.

I've made it pretty strict - it's limited to one bucket and folder inside it. Then, it's less likely authors mess up other content in our S3. They only need to provide a name of the glossary, not even filename. 

JSON files will be uploaded to `models-resources/glossary-resources/json/<name>.json`.

I'm working on another feature which will upload images and videos to `models-resources/glossary-resources/media/<filename>`.

There are some changes to schema validation. It will ensure that entries have unique "word" property and image and video URLs start with http(s).

kswenson commented on src/components/authoring/definition-editor.tsx:
```diff
@@ -14,15 +15,25 @@ interface IState {
+      image: this.props.initialDefinition && this.props.initialDefinition.image || "",
+      imageCaption: this.props.initialDefinition && this.props.initialDefinition.imageCaption || "",
+      video: this.props.initialDefinition && this.props.initialDefinition.video || "",
+      videoCaption: this.props.initialDefinition && this.props.initialDefinition.videoCaption || ""
```
A more compact alternative:
```javascript
definition: Object.assign({ word: "", definition: "", ...}, this.props.initialDefinition)
```

pjanik replied:
Thanks, way nicer!

pjanik commented:
I've addressed Kirks comments and added a few more changes.
E.g. replaced html-plugin with copy-plugin. html-plugin doesn't let you specify which JS files should be included on a given page (unless you add some plugin plugins, etc...). So, all the build products were included on every page, what didn't make sense. copy-plugin is straightforward and it's not more work in our case.

pjanik commented:
@knowuh, since this PR was still open, I've added image and video upload to this PR. Hopefully, it's not too big now. Commits should be descriptive if you prefer to review separately. Any chances you could take a look today? It would be helpful, as then I could address your comments tomorrow before the demo meeting. 

Re media upload, I've tried to find a compromise between handling errors and edge cases, and not creating something too complex. Dropzone will accept only selected file types, display errors, and limit upload to a single file. Potential errors are displayed at the bottom of the form.

Demo: 
http://glossary-plugin.concord.org/branch/160794190-authoring-s3/authoring.html

knowuh commented:
@pjanik  -- Sorry for not looking at this yet, I have been swamped.  

Maybe if you want this reviewed soon we can pull in @scytacki or @dougmartin to have a look?

knowuh commented:
Hey @kswenson -- just noticed that you have been reviewing this PR.   Want to keep going on it for me?

scytacki commented on src/utils/s3-helpers.tsx:
```diff
@@ -0,0 +1,70 @@
+      secretAccessKey: secretKey,
+    });
+    const key = `${S3_DIR_PREFIX}/${dir}/${filename}`;
+    s3.upload({
```
I think you could simplify this a little by using AWS's built in promise support. Here is the most official documentation I can found about it: https://aws.amazon.com/blogs/developer/support-for-promises-in-the-sdk/

Basically you can add `.promise()` to most every AWS api call.  Which means you can do something like:
`await s3.upload(...).promise();`

pjanik replied:
Thanks, that definitely looks cleaner and won't be much work.

scytacki commented on src/utils/s3-helpers.tsx:
```diff
@@ -0,0 +1,70 @@
+        resolve(`${CLOUDFRONT_URL}/${data.Key}`);
+      }
+    });
+    // Invalidate Cloudfront path too.
```
I personally would avoid cloudfront invalidations. They provide a false sense of fixing caching problems.  If you are are running into caching problems and you need to invalidate cloudfront, it would be better decrease the caching of the file. I'd guess for the json files it could be set to 0 seconds. And really long for the image and videos assuming they always have unique files names and you aren't updating them in place.

scytacki replied:
I looked at this more. And yes it would be better to set the caching behavior so it is short or 0 for the json.  Otherwise authors that don't clear their browser cache will be getting old versions of the glossary and not know why.

And I see the media files get uuid's prepended, so their caching time could be set to 1 year.

pjanik replied:
I've used the existing Cloudfront distribution (models-resources), so I can't really change its config I guess, as it would affect many projects.

The main problem is that if you save your changes and then load this JSON again, you won't see changes for a while. So, I think we shouldn't use caching for JSON, at least during authoring. Invalidations seemed to work fine for me, there was a tiny delay (~2 secons). But right, the browser cache could be more problematic (I had dev tools open when I was playing with it).

I can create a new distribution for the directory I'm using here and set custom caching rules or simply return direct S3 links to JSON. Direct S3 links seem simpler and less likely to get broken (no config in AWS involved). I'm wondering if there's any benefit of using Cloudfront with 0 caching time vs a direct S3 link? Cloudfront should hit S3 each time anyway. So, I'd rather use S3 links for JSONs and Cloudfront links for media.

scytacki replied:
I might be reading it wrong, but I think the models-resources cloudfront distribution will always prefer the cache headers sent by the origin.  So when the files are uploaded using s3.upload you can set the cache headers. You can see it listed as the CacheControl parameter here: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property
That header saved in S3 will override any settings in CloudFormation, and it should also be sent down to the user's browser.

I'm not sure if we get a benefit by using cloudfront with 0 caching. At least it would be uniform so all requests would be going through our concord.org domain. That might help with some schools that filter s3 domains. Actually one thing we get is automatic compression. If configured correctly the cloudfront distribution will gzip all json. I don't think S3 will do that by itself.

scytacki replied:
Just to be clear for other use cases. Even if the CF distribution customizes the Object Caching, it will not send this down to the browser. So using that to configure caching is not an alternative to setting the caching in S3.

scytacki replied:
I checked and the models-resources distribution is configured to compress objects automatically.

scytacki commented:
Other than the invalidations/caching this looks good. 
I'm not sure switching to AWS promises is worth it.

scytacki commented:
I just read this: https://www.rdegges.com/2018/please-stop-using-local-storage/

I don't think it is worth adding the extra layer to avoid the local storage of API keys at this time. But I think we need to be careful with this:, and encourage folks to use restricted AWS API keys.

pjanik commented:
@scytacki, I went ahead and switch between caching and not using Cloudfront or direct S3 links. I can switch to a custom Cloudfront distribution with different cache config for JSONs and media if you think or know that it will be beneficial. Also, switched to AWS promises and fixed a random, small bug I found.

scytacki commented:
I think it would be better to use the CacheControl param during upload instead of using S3 links to avoid caching. This will at least get us compression and have a consistent set of domains for the glossary files.

pjanik commented:
That could work, although there's one use case which doesn't work too well:
User provides some random name and clicks "Load" => "noSuchKey" page gets cached. Then, even if he saves this file correctly, it's impossible to load it (probably for some time). I know it's not very useful scenario, but it's easy to do (I did it many times while playing with this page / testing). If he saves the file first and there's never 404 / no such key returned, things work fine.

I've tried this custom error page: https://forums.aws.amazon.com/thread.jspa?threadID=220220
but it didn't help so far (unless it takes a while to update cloudfront config). I'll take a look tomorrow again.

pjanik commented:
> I've tried this custom error page: https://forums.aws.amazon.com/thread.jspa?threadID=220220
but it didn't help so far (unless it takes a while to update cloudfront config). I'll take a look tomorrow again.

Now it works fine, so I guess it took some time to update Cloudfront. I guess we could do that for all the buckets that we use for deployment, as I had this problem multiple times.
I've just pushed changes to use Cache-Control.

scytacki commented:
👍
